src/sentry/models/grouphistory.py

```diff
@@ -285,7 +286,7 @@ def record_group_history_from_activity_type(
 def record_group_history(
     group: Group,
     status: int,
-    actor: User | RpcUser | Team | None = None,
+    actor: User | RpcUser | Team | AnonymousUser | None = None,
```

Comment:
@asottile-sentry some endpoints do things like this:
https://github.com/getsentry/sentry/blob/d494c2da1f2866260a7fc8164d7440f12737a6d7/src/sentry/api/helpers/group_index/update.py#L174-175

This means that the user can be `AnonymousUser` in `record_group_history`, however, the function only handles these types: RpcUser, User or Team.

I assume that in reality we never pass anything like `AnonymousUser`, otherwise, we should see the error in Sentry.
https://sentry.my.sentry.io/organizations/sentry/issues/?project=2&query=record_group_history%20actor%20argument%20must%20be%20RPC&referrer=issue-list&statsPeriod=7d

I wonder if all the callers for this function should be passing down `acting_user` rather than `user`.

Reply:
yeah it's tricky -- generally I've been trying to either guard against anonymous explicitly in the view (even if auth has already probably done its job) -- `request.user` is kinda unfortunate in that there's some _implicit_ narrowing that can happen due to our auth scopes stuff but that isn't reflected directly in the type system

there's some hacks to define a "request subclass" that always has user -- but I feel like those are slightly worse than the check for anonymous in the view -- mostly because they aren't checked for correctness at all and sorta subvert the type system

I think the ultimate answer to this is "don't pass around the request object" but instead pull out the parts you need from it and pass those along.  this also makes it easier to test (and you don't have to think "is this a drf Request? a django HttpRequest? requests Request?" you just have the plain values to operate on!)